File: src/main/java/eu/monsbot/turbokarts/vehiclelogic/karts/vehicles/AbstractTKVehicle.java
package eu.monsbot.turbokarts.vehiclelogic.karts.vehicles;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.bukkit.Color;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.LeatherArmorMeta;
import org.bukkit.inventory.meta.PotionMeta;
import org.bukkit.util.Transformation;
import org.joml.AxisAngle4f;
import org.joml.Vector3f;

public abstract class AbstractTKVehicle {

    /** -- physics -- **/
    int maxSpeed;           //  kb/h = km/h
    float acceleration;     //  b/s²
    float deceleration;     //  b/s²
    int mass;               //  g
    float grip;             //  μ on concrete
    float handlingRadius;   //  b
    Vector3f hitbox;        //  bxbxb xyz

    /** -- visuals -- **/
    // name
    private String name;
    private String brand;
    // primary
    Material primaryMaterial;
    int      primaryCustomModelData;
    int      primaryColor;
    Vector3f primaryTranslation;
    Vector3f primarySize;
    AxisAngle4f primaryLeftRotation;
    AxisAngle4f primaryRightRotation;
    // accent1
    Material accent1Material;
    int      accent1CustomModelData;
    int      accent1Color;
    Vector3f accent1Translation;
    Vector3f accent1Size;
    AxisAngle4f accent1LeftRotation;
    AxisAngle4f accent1RightRotation;

    // accent2
    Material accent2Material;
    int      accent2CustomModelData;
    int      accent2Color;
    Vector3f accent2Translation;
    Vector3f accent2Size;
    AxisAngle4f accent2LeftRotation;
    AxisAngle4f accent2RightRotation;

    // steering wheel
    Material steeringWHeelMaterial;
    int      steeringWHeelCustomModelData;
    int      steeringWHeelColor;
    Vector3f steeringWHeelTranslation;
    Vector3f steeringWHeelSize;
    AxisAngle4f steeringWheelLeftRotation;
    AxisAngle4f steeringWheelRightRotation;

    // dummy
    Vector3f headTranslation;
    AxisAngle4f headLeftRotation;
    AxisAngle4f headRightRotation;

    Vector3f bodyTranslation;
    AxisAngle4f bodyLeftRotation;
    AxisAngle4f bodyRightRotation;

    Vector3f leftArmTranslation;
    AxisAngle4f leftArmLeftRotation;
    AxisAngle4f leftArmRightRotation;
    Vector3f rightArmTranslation;
    AxisAngle4f rightArmLeftRotation;
    AxisAngle4f rightArmRightRotation;

    Vector3f leftLegTranslation;
    AxisAngle4f leftLegLeftRotation;
    AxisAngle4f leftLegRightRotation;
    Vector3f rightLegTranslation;
    AxisAngle4f rightLegLeftRotation;
    AxisAngle4f rightLegRightRotation;

    public Transformation getBasePrimaryTransformation() {
        return new Transformation(
                primaryTranslation,
                primaryLeftRotation,
                primarySize,
                primaryRightRotation
        );
    }


    /**
     *  The json needed to write down a vehicle model ({@link AbstractTKVehicle}):
     * [ physics ]
     *  - max speed
     *  - acceleration
     *  - mass
     *  - hitbox
     *  - handling
     *  - grip
     * [ visual ]
     *  - name
     *  - array for the values for the body displays
     *  - dummy pose
     *  - wheel item display properties
     *  - wheel position
     *  - wheel rotation direction & axis (Vector)
     *  - steering wheel position
     *  META: brand
     **/

    public String toJson() {
        Gson gson = new Gson();
        return gson.toJson(this);
    }

    public static AbstractTKVehicle fromJson(String json) {
        Gson gson = new Gson();
        return gson.fromJson(json, new TypeToken<AbstractTKVehicle>(){}.getType());
    }

    private ItemStack getItem(Material material, int customModelData, int color) {
        ItemStack itemStack = new ItemStack(material);
        ItemMeta itemMeta = itemStack.getItemMeta();
        itemMeta.setCustomModelData(customModelData);
        if (itemMeta instanceof PotionMeta && color > -1) {
            PotionMeta potionMeta = (PotionMeta) itemMeta;
            potionMeta.setColor(Color.fromRGB(color));

        } else if (itemMeta instanceof LeatherArmorMeta && color > -1) {
            LeatherArmorMeta leatherArmorMeta = (LeatherArmorMeta) itemMeta;
            leatherArmorMeta.setColor(Color.fromRGB(color));
        }
        itemStack.setItemMeta(itemMeta);
        return itemStack;
    }

    public ItemStack getBasePrimaryItem() {
        return getItem(primaryMaterial, primaryCustomModelData, primaryColor);
    }

    public Transformation getBaseAccent1Transformation() {
        return new Transformation(
                accent1Translation,
                accent1LeftRotation,
                accent1Size,
                accent1RightRotation
        );
    }

    public ItemStack getBaseAccent1Item() {
        return getItem(accent1Material, accent1CustomModelData, accent1Color);
    }

    public Transformation getBaseAccent2Transformation() {
        return new Transformation(
                accent2Translation,
                accent2LeftRotation,
                accent2Size,
                accent2RightRotation
        );
    }

    public ItemStack getBaseAccent2Item() {
        return getItem(accent2Material, accent2CustomModelData, accent2Color);
    }

    public Transformation getSteeringWheelTransformation() {
        return new Transformation(
                headTranslation,
                headLeftRotation,
                new Vector3f(1, 1, 1),
                headRightRotation
        );
    }

    public ItemStack getSteeringWheelItem() {
        return getItem(steeringWHeelMaterial, steeringWHeelCustomModelData, steeringWHeelColor);
    }
}

File: src/main/kotlin/eu/monsbot/turbokarts/vehiclelogic/karts/vehicles/AbstractTKVehicle.kt
package eu.monsbot.turbokarts.vehiclelogic.karts.vehicles

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.bukkit.Color
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.LeatherArmorMeta
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.util.Transformation
import org.joml.AxisAngle4f
import org.joml.Vector3f

abstract class AbstractTKVehicle {
    /** -- physics --  */
    var maxSpeed: Int = 0 //  kb/h = km/h
    var acceleration: Float = 0f //  b/s²
    var deceleration: Float = 0f //  b/s²
    var mass: Int = 0 //  g
    var grip: Float = 0f //  μ on concrete
    var handlingRadius: Float = 0f //  b
    var hitbox: Vector3f? = null //  bxbxb xyz

    /** -- visuals --  */ // name
    private val name: String? = null
    private val brand: String? = null

    // primary
    var primaryMaterial: Material? = null
    var primaryCustomModelData: Int = 0
    var primaryColor: Int = 0
    var primaryTranslation: Vector3f? = null
    var primarySize: Vector3f? = null
    var primaryLeftRotation: AxisAngle4f? = null
    var primaryRightRotation: AxisAngle4f? = null

    // accent1
    var accent1Material: Material? = null
    var accent1CustomModelData: Int = 0
    var accent1Color: Int = 0
    var accent1Translation: Vector3f? = null
    var accent1Size: Vector3f? = null
    var accent1LeftRotation: AxisAngle4f? = null
    var accent1RightRotation: AxisAngle4f? = null

    // accent2
    var accent2Material: Material? = null
    var accent2CustomModelData: Int = 0
    var accent2Color: Int = 0
    var accent2Translation: Vector3f? = null
    var accent2Size: Vector3f? = null
    var accent2LeftRotation: AxisAngle4f? = null
    var accent2RightRotation: AxisAngle4f? = null

    // steering wheel
    var steeringWHeelMaterial: Material? = null
    var steeringWHeelCustomModelData: Int = 0
    var steeringWHeelColor: Int = 0
    var steeringWHeelTranslation: Vector3f? = null
    var steeringWHeelSize: Vector3f? = null
    var steeringWheelLeftRotation: AxisAngle4f? = null
    var steeringWheelRightRotation: AxisAngle4f? = null

    // dummy
    var headTranslation: Vector3f? = null
    var headLeftRotation: AxisAngle4f? = null
    var headRightRotation: AxisAngle4f? = null

    var bodyTranslation: Vector3f? = null
    var bodyLeftRotation: AxisAngle4f? = null
    var bodyRightRotation: AxisAngle4f? = null

    var leftArmTranslation: Vector3f? = null
    var leftArmLeftRotation: AxisAngle4f? = null
    var leftArmRightRotation: AxisAngle4f? = null
    var rightArmTranslation: Vector3f? = null
    var rightArmLeftRotation: AxisAngle4f? = null
    var rightArmRightRotation: AxisAngle4f? = null

    var leftLegTranslation: Vector3f? = null
    var leftLegLeftRotation: AxisAngle4f? = null
    var leftLegRightRotation: AxisAngle4f? = null
    var rightLegTranslation: Vector3f? = null
    var rightLegLeftRotation: AxisAngle4f? = null
    var rightLegRightRotation: AxisAngle4f? = null

    val basePrimaryTransformation: Transformation
        get() = Transformation(
                primaryTranslation!!,
                primaryLeftRotation!!,
                primarySize!!,
                primaryRightRotation!!
        )


    /**
     * The json needed to write down a vehicle model ([AbstractTKVehicle]):
     * [ physics ]
     * - max speed
     * - acceleration
     * - mass
     * - hitbox
     * - handling
     * - grip
     * [ visual ]
     * - name
     * - array for the values for the body displays
     * - dummy pose
     * - wheel item display properties
     * - wheel position
     * - wheel rotation direction & axis (Vector)
     * - steering wheel position
     * META: brand
     */
    fun toJson(): String {
        val gson = Gson()
        return gson.toJson(this)
    }

    private fun getItem(material: Material?, customModelData: Int, color: Int): ItemStack {
        val itemStack = ItemStack(material!!)
        val itemMeta = itemStack.itemMeta
        itemMeta.setCustomModelData(customModelData)
        if (itemMeta is PotionMeta && color > -1) {
            itemMeta.color = Color.fromRGB(color)
        } else if (itemMeta is LeatherArmorMeta && color > -1) {
            itemMeta.setColor(Color.fromRGB(color))
        }
        itemStack.setItemMeta(itemMeta)
        return itemStack
    }

    val basePrimaryItem: ItemStack
        get() = getItem(primaryMaterial, primaryCustomModelData, primaryColor)

    val baseAccent1Transformation: Transformation
        get() = Transformation(
                accent1Translation!!,
                accent1LeftRotation!!,
                accent1Size!!,
                accent1RightRotation!!
        )

    val baseAccent1Item: ItemStack
        get() = getItem(accent1Material, accent1CustomModelData, accent1Color)

    val baseAccent2Transformation: Transformation
        get() = Transformation(
                accent2Translation!!,
                accent2LeftRotation!!,
                accent2Size!!,
                accent2RightRotation!!
        )

    val baseAccent2Item: ItemStack
        get() = getItem(accent2Material, accent2CustomModelData, accent2Color)

    val steeringWheelTransformation: Transformation
        get() = Transformation(
                headTranslation!!,
                headLeftRotation!!,
                Vector3f(1f, 1f, 1f),
                headRightRotation!!
        )


    val steeringWheelItem: ItemStack
        get() = getItem(steeringWHeelMaterial, steeringWHeelCustomModelData, steeringWHeelColor)

    companion object {
        fun fromJson(json: String?): AbstractTKVehicle {
            val gson = Gson()
            return gson.fromJson(json, object : TypeToken<AbstractTKVehicle?>() {}.type)
        }
    }
}